src/main/java/com/linecorp/armeria/server/metrics/ServiceMetricDecorator.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2015 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.metrics;
+
+import java.util.function.Supplier;
+
+import com.linecorp.armeria.common.ServiceInvocationContext;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.server.DecoratingServiceCodec;
+import com.linecorp.armeria.server.RequestTimeoutException;
+import com.linecorp.armeria.server.ServiceCodec;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.util.AttributeKey;
+import io.netty.util.concurrent.Promise;
+
+/**
+ * Decorator to collect service metrics.
+ *
+ * This class is expected to be used with other ServiceMetricConsumer
+ */
+public class ServiceMetricDecorator extends DecoratingServiceCodec {
+    private static final AttributeKey<Metrics> METRICS =
+            AttributeKey.valueOf(ServiceMetricDecorator.class, "METRICS");
+
+    private final ServiceMetricConsumer metricConsumer;
+
+    /**
+     * Creates a new instance that decorates the specified {@link ServiceCodec}.
```

Comment:
the specified ... with the specified `{@code ServiceMetricConsumer}` (or `{@code consumer}`)